### ISSUE TITLE
[discord-game-sdk] build failure with MSVC

### ISSUE THREAD

**Issue by TheMaverickProgrammer** [microsoft/vcpkg#29306]
**Host Environment**

- OS: Windows
- Compiler: Microsoft (R) C/C++ Optimizing Compiler Version 19.29.30137 for x86

**To Reproduce**

`./vcpkg install discord-game-sdk:x64-windows`

**Failure logs**
```
LNK2038	mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in main.obj discord_game_sdk_cpp.lib(achievement_manager.cpp.obj)	1

LNK2038	mismatch detected for 'RuntimeLibrary': value 'MD_DynamicRelease' doesn't match value 'MDd_DynamicDebug' in main.obj discord_game_sdk_cpp.lib(achievement_manager.cpp.obj)	1	

LNK4098	defaultlib 'MSVCRT' conflicts with use of other libs; use /NODEFAULTLIB:library

22 Mismatches detected
```

**Additional context**

This problem occurs because the current port does not properly select the correct libs based on the build type.
In debug they show up like this:

```
X:\Big_Code\vcpkg\installed\x64-windows\debug\lib\fluidsynth.lib
X:\Big_Code\vcpkg\installed\x64-windows\debug\lib\PocoNetd.lib
X:\Big_Code\vcpkg\installed\x64-windows\debug\lib\PocoJSONd.lib
X:\Big_Code\vcpkg\installed\x64-windows\lib\discord_game_sdk.lib
X:\Big_Code\vcpkg\installed\x64-windows\lib\discord_game_sdk_cpp.lib
X:\Big_Code\vcpkg\installed\x64-windows\debug\lib\lua.lib
```

When I manually change the game sdk path to
```
X:\Big_Code\vcpkg\installed\x64-windows\debug\lib\discord_game_sdk.lib
X:\Big_Code\vcpkg\installed\x64-windows\debug\lib\discord_game_sdk_cpp.lib
```

Everything works as expected. The port needs to have the correct cmake configuration rules to detect if needs debug or release linkage.
Note that while this problem is very similar to [26342](https://github.com/microsoft/vcpkg/issues/26342), my project has dynamic linkages and I am not writing my own triplet to fix this issue. Manually fixing the issue will prevent us from doing CI/CD.

**Comment by JonLiu1993** [comment]
@TheMaverickProgrammer, Thanks for posting this issue, I can install `discord-game-sdk:x64-windows` successfully on my machine with windows11 vs2022:
````
PS F:\test\vcpkg> ./vcpkg install discord-game-sdk:x64-windows --editable
Computing installation plan...
The following packages will be built and installed:
    discord-game-sdk[core]:x64-windows -> 2.5.6#4
Restored 0 package(s) from C:\Users\vzhli17\AppData\Local\vcpkg\archives in 152.3 us. Use --debug to see more details.
Installing 1/1 discord-game-sdk:x64-windows...
Building discord-game-sdk[core]:x64-windows...
-- Using cached discord_game_sdk.zip.
-- Extracting source F:/test/vcpkg/downloads/discord_game_sdk.zip
-- Applying patch include-cstdint.patch
-- Using source at F:/test/vcpkg/buildtrees/discord-game-sdk/src/d_game_sdk-19a6f72d70
-- Installing: F:/test/vcpkg/packages/discord-game-sdk_x64-windows/bin/discord_game_sdk.dll
-- Installing: F:/test/vcpkg/packages/discord-game-sdk_x64-windows/debug/bin/discord_game_sdk.dll
-- Installing: F:/test/vcpkg/packages/discord-game-sdk_x64-windows/lib/discord_game_sdk.lib
-- Installing: F:/test/vcpkg/packages/discord-game-sdk_x64-windows/debug/lib/discord_game_sdk.lib
-- Found external ninja('1.11.0').
-- Configuring x64-windows
-- Building x64-windows-dbg
-- Building x64-windows-rel
-- Installing: F:/test/vcpkg/packages/discord-game-sdk_x64-windows/share/discord-game-sdk/copyright
-- Installing: F:/test/vcpkg/packages/discord-game-sdk_x64-windows/share/discord-game-sdk/usage
-- Performing post-build validation
Elapsed time to handle discord-game-sdk:x64-windows: 14.11 s
Total install time: 14.14 s
The package discord-game-sdk does not provides CMake integration:

    find_library(GAME_SDK discord_game_sdk)
    target_link_libraries(<my_target> PRIVATE ${GAME_SDK})

    find_path(GAME_SDK_INCLUDE discord-game-sdk/discord.h)
    target_include_directories(<my_target> PRIVATE ${GAME_SDK_INCLUDE})

The C++ SDK source code is compiled in a static library:

    find_library(CPP_GAME_SDK discord_game_sdk_cpp)
    target_link_libraries(<my_target> PRIVATE ${CPP_GAME_SDK})
````

Did your error occur when installing discord-game-sdk or when using discord-game-sdk?

**Comment by TheMaverickProgrammer** [comment]
> @TheMaverickProgrammer, Thanks for posting this issue, I can install `discord-game-sdk:x64-windows` successfully on my machine with windows11 vs2022:
> 
> ```
> PS F:\test\vcpkg> ./vcpkg install discord-game-sdk:x64-windows --editable
> Computing installation plan...
> The following packages will be built and installed:
>     discord-game-sdk[core]:x64-windows -> 2.5.6#4
> Restored 0 package(s) from C:\Users\vzhli17\AppData\Local\vcpkg\archives in 152.3 us. Use --debug to see more details.
> Installing 1/1 discord-game-sdk:x64-windows...
> Building discord-game-sdk[core]:x64-windows...
> -- Using cached discord_game_sdk.zip.
> -- Extracting source F:/test/vcpkg/downloads/discord_game_sdk.zip
> -- Applying patch include-cstdint.patch
> -- Using source at F:/test/vcpkg/buildtrees/discord-game-sdk/src/d_game_sdk-19a6f72d70
> -- Installing: F:/test/vcpkg/packages/discord-game-sdk_x64-windows/bin/discord_game_sdk.dll
> -- Installing: F:/test/vcpkg/packages/discord-game-sdk_x64-windows/debug/bin/discord_game_sdk.dll
> -- Installing: F:/test/vcpkg/packages/discord-game-sdk_x64-windows/lib/discord_game_sdk.lib
> -- Installing: F:/test/vcpkg/packages/discord-game-sdk_x64-windows/debug/lib/discord_game_sdk.lib
> -- Found external ninja('1.11.0').
> -- Configuring x64-windows
> -- Building x64-windows-dbg
> -- Building x64-windows-rel
> -- Installing: F:/test/vcpkg/packages/discord-game-sdk_x64-windows/share/discord-game-sdk/copyright
> -- Installing: F:/test/vcpkg/packages/discord-game-sdk_x64-windows/share/discord-game-sdk/usage
> -- Performing post-build validation
> Elapsed time to handle discord-game-sdk:x64-windows: 14.11 s
> Total install time: 14.14 s
> The package discord-game-sdk does not provides CMake integration:
> 
>     find_library(GAME_SDK discord_game_sdk)
>     target_link_libraries(<my_target> PRIVATE ${GAME_SDK})
> 
>     find_path(GAME_SDK_INCLUDE discord-game-sdk/discord.h)
>     target_include_directories(<my_target> PRIVATE ${GAME_SDK_INCLUDE})
> 
> The C++ SDK source code is compiled in a static library:
> 
>     find_library(CPP_GAME_SDK discord_game_sdk_cpp)
>     target_link_libraries(<my_target> PRIVATE ${CPP_GAME_SDK})
> ```
> 
> Did your error occur when installing discord-game-sdk or when using discord-game-sdk?

This is through cmake installing via with github workflow

**Comment by TheMaverickProgrammer** [comment]
> @TheMaverickProgrammer, I investigated and found that the files under the debug\lib and \lib paths after successfully installing discord-game-sdk are the same:
> 
> ![image](https://user-images.githubusercontent.com/63675417/225859167-b0a10c38-c7ba-4ba8-8fd5-748805d9708c.png)
> 
> ![image](https://user-images.githubusercontent.com/63675417/225859283-eebe9eef-2be9-441f-a94b-5bbc6da9c127.png)
> 
> Are you using the latest vcpkg?

Your `discord_game_sdk_cpp.lib` is clearly not the same in file size. Which according to my original error is the problem;

```
discord_game_sdk_cpp.lib(achievement_manager.cpp.obj)
```

**Comment by dg0yt** [comment]
@TheMaverickProgrammer The "Steps to reproduce the behavior" could be understood as a problem with *building* `discord-game-sdk:x64-windows`, but actually you face problems *using* `discord-game-sdk:x64-windows`.
The printed usage instructions are for single-config generators. AFAICT they would cause the reported problem for the debug variant of a multi-config cmake project (e.g. using VS generators).

**Comment by TheMaverickProgrammer** [comment]
> 

Do you have a solution? All other vcpkgs I use correctly link to the correct symbols (debug/release).

**Comment by dg0yt** [comment]
So you really want to use multi-config? The key is to use two calls to `find_library` and then continue with `select_library_configurations` or setting up an imported target.

**Comment by TheMaverickProgrammer** [comment]
> So you really want to use multi-config? The key is to use two calls to `find_library` and then continue with `select_library_configurations` or setting up an imported target.

Any idea why all my other vcpkg libraries are correctly configured without doing this extra setup with `select_library_configurations`?

**Comment by dg0yt** [comment]
It is not possible to say something about *all your* other libraries. But for those mentioned:

X:\Big_Code\vcpkg\installed\x64-windows\debug\lib\fluidsynth.lib
Port fluidsynth provides CMake config with an imported target.

X:\Big_Code\vcpkg\installed\x64-windows\debug\lib\PocoNetd.lib
X:\Big_Code\vcpkg\installed\x64-windows\debug\lib\PocoJSONd.lib
Port poco provides CMake config with imported targets.

X:\Big_Code\vcpkg\installed\x64-windows\debug\lib\lua.lib
Port lua modifies  CMake's regular find module output variables, leveraging unofficial CMake config with an imported target.

**Comment by TheMaverickProgrammer** [comment]
> It is not possible to say something about _all your_ other libraries. But for those mentioned:
> 
> X:\Big_Code\vcpkg\installed\x64-windows\debug\lib\fluidsynth.lib Port fluidsynth provides CMake config with an imported target.
> 
> X:\Big_Code\vcpkg\installed\x64-windows\debug\lib\PocoNetd.lib X:\Big_Code\vcpkg\installed\x64-windows\debug\lib\PocoJSONd.lib Port poco provides CMake config with imported targets.
> 
> X:\Big_Code\vcpkg\installed\x64-windows\debug\lib\lua.lib Port lua modifies CMake's regular find module output variables, leveraging unofficial CMake config with an imported target.

Thank you for pointing this out. It was my understanding from all of my experience so far with vcpkg ports that the imported target should be correctly configured otherwise it was a problem with the port's build scripts. I'll try your suggestion and see if it solves my issue. 

In the mean time, what do you mean by (and I'm paraphrasing) me "setting up my own imported target"? Is that not the same thing as writing my own port? Because that is not ideal and is very time consuming. I'd prefer using the extra CMake steps to configure the target correctly and build as I'd expect it to.

**Comment by dg0yt** [comment]
You are not meant to modify the port. You simply need to do elaborate the `find_*` steps in your own project. That's almost the whole magic behind Find modules. Modern Find modules add libraries of type `UNKNOWN IMPORTED` and set per-config properties from the results of the `find_*` calls. 

The key strength of exported CMake config is that it can easily capture not only install locations but also configuration options and transitive usage requirements. But this port doesn't declare any dependencies.

In generally it is upstream's task to provide official CMake config.

Last not least, this is the code for settung up variables (untested):
~~~cmake
find_library(GAME_SDK_LIBRARY_RELEASE NAMES discord_game_sdk PATHS "{VCPKG_INSTALLED_DIR}/${VCPKG_TARGET_TRIPLET}/lib" NO_DEFAULT_PATH)
find_library(GAME_SDK_LIBRARY_DEBUG NAMES discord_game_sdk PATHS "{VCPKG_INSTALLED_DIR}/${VCPKG_TARGET_TRIPLET}/debug/lib" NO_DEFAULT_PATH)
include(SelectLibraryConfigurations)
select_library_configurations(GAME_SDK)
target_link_libraries(<my_target> PRIVATE ${GAME_SDK_LIBRARIES})
~~~
(I would like to avoid the `PATHS` option with the  `VCPKG_INSTALLED_DIR` variable by using `PATH_SUFFIXES` instead, but this should really be tested first, because it interacts in a more complex way with standard CMake variables.)

**Comment by TheMaverickProgrammer** [comment]
> In generally it is upstream's task to provide official CMake config.

So you are affirming this **is** an issue but you have provided a manual solution. 

Who is upstream in this case? VCPKG port maintainers or Discord Game SDK owners? If VPKG, we're in the right thread and we can wait for the issue to be pickup and resolved. If it's the Discord Game SDK owners, I can send this with my findings to their github.

**Comment by dg0yt** [comment]
Thanks for asking again ... Looking more closely, I have to rectify some statements.
- vcpkg doesn't build the `discord_game_sdk` lib but just copies upstreams artifacts into release and debug directories. 
  So for this (C ?) lib, there is no difference between debug and release.
- However, vcpkg builds the `discord_game_sdk_cpp` lib. This C++ library must be used correctly for debug vs. release, similar to what I presented in the cmake snippet.
  (And this is the lib where you saw the error.)
- The build of the C++ lib is done by a CMake project which comes with the vcpkg port. So there is no point in asking upstream to fix it, unless they would be willing to add similar CMake.

Given that upstream does provide *any* build system for the C++ lib, and no official lib name, providing this C++ library already is vcpkg lock-in, and adding unofficial CMake config directly to the vcpkg port would be the right thing to do.
(This being said, one could submit a simple CMake build system to the Discord Game SDK owners, but it is unclear if this is what they want and which additional question they would raise.)

**Comment by TheMaverickProgrammer** [comment]
Sorry your sentence structure doesn't make sense and it is hard to follow. 

Anyway, I tried your suggestion and it includes both debug and release libraries into the project for both debug and release build configurations.

It's definitely promising that it did find the `debug` library.

**Comment by TheMaverickProgrammer** [comment]
Update, I changed 
```cmake
target_link_libraries(<my_target> PRIVATE ${GAME_SDK_LIBRARIES})
```
to just 

```cmake
target_link_libraries(<my_target> PRIVATE ${GAME_SDK_LIBRARY})
```

and I see the correct paths now. I'll double check with another clean build. 👍

**Comment by TheMaverickProgrammer** [comment]
Here's my final cmake to make this work right. I confirmed with a full clean build on both debug and release:

```cmake
if(MY_PROJECT_DISCORD_SUPPORT)
	# Headers
	find_library(DISCORD_GAME_SDK discord_game_sdk)
	target_link_libraries(<my_target> ${DISCORD_GAME_SDK})
	find_path(DISCORD_GAME_SDK_INCLUDE discord-game-sdk/discord.h)
	target_include_directories(<my_target> PRIVATE ${DISCORD_GAME_SDK_INCLUDE})

	# The C++ SDK source code is compiled to a static library
	# 
	# Fix for vcpkg port to select correct symbols 03/19/2023
	# https://github.com/microsoft/vcpkg/issues/29306#issuecomment-1475303879
	find_library(DISCORD_CPP_GAME_SDK_LIBRARY_RELEASE NAMES discord_game_sdk_cpp
		PATHS "${VCPKG_INSTALLED_DIR}" 
		PATH_SUFFIXES "${VCPKG_TARGET_TRIPLET}/lib" NO_DEFAULT_PATH)
	find_library(DISCORD_CPP_GAME_SDK_LIBRARY_DEBUG NAMES discord_game_sdk_cpp
		PATHS "${VCPKG_INSTALLED_DIR}"
		PATH_SUFFIXES "${VCPKG_TARGET_TRIPLET}/debug/lib" NO_DEFAULT_PATH)
	
	include(SelectLibraryConfigurations)
	select_library_configurations(DISCORD_CPP_GAME_SDK)
	
	target_link_libraries(<my_target> ${DISCORD_CPP_GAME_SDK_LIBRARY})
endif()
```
